### ISSUE TITLE
now() is evaluated only once per survey

### ISSUE THREAD

**Issue by grzesiek2010** [opendatakit/javarosa#199]
I'm in the middle of investigating older issues in the collect repository and I found these tickets:
https://github.com/opendatakit/collect/issues/140
https://github.com/opendatakit/collect/issues/169

They have "needs reproduction" label but reproduction is easy. I think it'a an issue on the Javarosa side. I made an initial investigation using this form:
[now.txt](https://github.com/opendatakit/javarosa/files/1375828/now.txt)

The attached form contains two functions: `sum() `and `now()`. `sum()` is called everytime I navigate through the form but now only once (just after opening the form).

Why do we call `now()` only once?

**Comment by lognaturel** [comment]
This seems like a bug, I can't think of any reason why this would be the desired behavior. Fixing it would enable lightweight timestamping as described in opendatakit/xforms-spec#123.

**Comment by lognaturel** [comment]
I did some exploration with 
[multiple-now.xml.txt](https://github.com/opendatakit/javarosa/files/1475947/multiple-now.xml.txt) to see what was happening with multiple calls to `now()` in the same form. I confirm the behavior that @grzesiek2010 describes and that it should not be this way.

This also led to some questions:
- Why does simply displaying `now()` result in a date without a time? Time is included because I can see it change when I wrap the call in `decimal-date-time`.
- Similarly, why can't I pull out the times using `format-date`? I seem to remember this coming up recently but I can't find the forum post or issue.
- What date/time format comes out of `decimal-date-time`? For example, I just got 17485.793219953703. Looks like the 17485 is days from epoch (Jan 1 1970). It probably uses the Excel format? http://excelsemipro.com/2010/08/date-and-time-calculation-in-excel/ Is this documented anywhere? https://forum.opendatakit.org/t/calculating-duration/2959/3 suggests this is the case but that time calculations have problems. Can we identify those?
- I tried `today()` to make sure it behaves as documented [here](https://opendatakit.github.io/xforms-spec/#fn:today) and it does

**Comment by lognaturel** [comment]
An additional note -- I got `17485.793219953703` at around 11am so I think the time given is in UTC rather than local time.

**Comment by lognaturel** [comment]
Now that I think about it, I think this is consistent with how other functions are evaluated. They are first evaluated on form entry and then reevaluated as related questions are updated. It has nothing to do with when the question is actually visited by the user.

**Comment by mdudzinski** [comment]
@lognaturel Yeah, you're right. How about introducing a function that returns a timestamp but is takes a node reference as an argument? It would be recalculated everytime the referenced node has been updated. That's the simplest solution I can think of.

**Comment by mdudzinski** [comment]
@ln It resutls with date instead of date time because of the following snippet from `Recalculate#wrapData(Object val, int dataType)`.
```
        } else if (val instanceof Date) {
            if ( dataType == Constants.DATATYPE_DATE_TIME) {
                return new DateTimeData((Date)val);
            }
            else if ( dataType == Constants.DATATYPE_TIME ) {
                return new TimeData((Date)val);
            } else {
                return new DateData((Date)val);
            }
```
It checks the data type of the node (`DATATYPE_TEXT` in your scenario). It this method, it's never tested against the `DATATYPE_TEXT` constant so the execution falls to the above check. As a result, a`DateData` instance is returned. Can be easily fixed I think :)

**Comment by lognaturel** [comment]
Will file a separate issue for the data type. 

I left an inline comment about the value change check in `commitAnswer`.

This is converging back with https://github.com/opendatakit/xforms-spec/issues/123 now that we are all on the same page about the behavior being consistent with all function evaluation. We do need some kind of new mechanism after all. 

@mdudzinski your idea of introducing a new function that takes in a node and reports a time of edit is an interesting one. I was hoping I could wrap it in `once()` to only get the first edit but of course it is evaluated immediately on form launch. 

What you have implemented would be called something like `last-time-edited(node)`. What people probably have in mind is more like `first-time-edited(node)` but actually I think the last edit is more insightful since it does give a sense of whether an enumerator jumped around.

I've passed that on to https://github.com/opendatakit/xforms-spec/issues/123 to see whether this is a spec addition we could agree on.

**Comment by mdudzinski** [comment]
@lognaturel I think both are possible, please check my comment [here](https://github.com/opendatakit/xforms-spec/issues/123#issuecomment-345674124).

**Comment by lognaturel** [comment]
Awesome, @mdudzinski!

@grzesiek2010 does all this business about why it's not a bug after all make sense? It points to a need for better docs about evaluation behavior which I've filed at https://github.com/opendatakit/docs/issues/213

**Comment by mdudzinski** [comment]
@lognaturel Arghh actually I was wrong ; \ won't be so easy with how dag works currently, sorry!

**Comment by lognaturel** [comment]
@mdudzinski You're talking about `first-time-edited(node)` being hard because everything is evaluated on form launch, right? That's ok, it's what I figured! Like I said, I do think the last time edited is more insightful anyway.

But all this is now making me worried about the `last-time-edited(node)` idea. With the simple implementation proposed would it just always have the finalize time? I haven't had a chance to check this yet but I suspect that's the case. I think everything is evaluated at form launch and form finalization.

**Comment by mdudzinski** [comment]
@lognaturel No, it won't run again at finalize if there is the fix in the `commitAnswer` method we discussed in 8ae88bd. But if you open the previously saved form, it will evaluate the `now()` functions anyway. So the only solution I've got for now is described [here](https://github.com/opendatakit/xforms-spec/issues/123#issuecomment-345944443).

**Comment by lognaturel** [comment]
As discussed above, this is a side effect of evaluation rules being consistently applied. https://github.com/opendatakit/xforms-spec/issues/123 explores some spec additions that could provide the desired behavior.
